File: aerox.Runtime.Graphics/Descriptors/DescriptorAllocator.cs
using TerraFX.Interop.Vulkan;
using static TerraFX.Interop.Vulkan.Vulkan;

namespace aerox.Runtime.Graphics.Descriptors;

public class DescriptorAllocator : Disposable
{
    private readonly VkDevice _device;
    private readonly List<DescriptorPool> _fullPools = new();
    private readonly PoolSizeRatio[] _ratios;
    private readonly List<DescriptorPool> _readyPools = new();
    private uint _setsPerPool;

    public DescriptorAllocator(uint maxSets, PoolSizeRatio[] poolRatios)
    {
        _ratios = poolRatios;
        _setsPerPool = maxSets;
        _device = Runtime.Instance.GetModule<GraphicsModule>().GetDevice();
    }

    protected override void OnDispose(bool isManual)
    {
        DestroyPools();
    }

    public DescriptorSet Allocate(VkDescriptorSetLayout layout)
    {
        var targetPool = GetPool();
        DescriptorSet set;
        try
        {
            set = targetPool.Allocate(layout);
        }
        catch (Exception)
        {
            _fullPools.Add(targetPool);
            targetPool = GetPool();
            set = targetPool.Allocate(layout);
        }

        _readyPools.Add(targetPool);

        return set;
    }

    public void DestroyPools()
    {
        foreach (var pool in _fullPools) pool.Dispose();

        _fullPools.Clear();


        foreach (var pool in _readyPools) pool.Dispose();

        _readyPools.Clear();
    }

    public void ClearPools()
    {
        foreach (var pool in _readyPools) pool.Reset();

        foreach (var pool in _fullPools)
        {
            pool.Reset();
            _readyPools.Add(pool);
        }

        _fullPools.Clear();
    }

    private DescriptorPool GetPool()
    {
        DescriptorPool pool;
        if (_readyPools.Count > 0)
        {
            pool = _readyPools.Last();
            _readyPools.RemoveAt(_readyPools.Count - 1);
        }
        else
        {
            pool = CreatePool(_setsPerPool, _ratios);
            _setsPerPool = (uint)(_setsPerPool * 1.5);
            if (_setsPerPool > 4092) _setsPerPool = 4092;
        }

        return pool;
    }

    private unsafe DescriptorPool CreatePool(uint setCount, PoolSizeRatio[] poolRatios)
    {
        var poolSizes = poolRatios.Select(r =>
        {
            VkDescriptorPoolSize size = default;
            size.type = r.type;
            size.descriptorCount = (uint)(r.ratio * setCount);
            return size;
        }).ToArray();


        var poolInfo = new VkDescriptorPoolCreateInfo
        {
            sType = VkStructureType.VK_STRUCTURE_TYPE_DESCRIPTOR_POOL_CREATE_INFO,
            flags = VkDescriptorPoolCreateFlags.VK_DESCRIPTOR_POOL_CREATE_UPDATE_AFTER_BIND_BIT,
            maxSets = setCount
        };
        fixed (VkDescriptorPoolSize* pPoolSizes = poolSizes)
        {
            poolInfo.pPoolSizes = pPoolSizes;
            poolInfo.poolSizeCount = (uint)poolSizes.Length;
        }

        VkDescriptorPool pool;
        vkCreateDescriptorPool(_device, &poolInfo, null, &pool);
        return new DescriptorPool(_device, pool);
    }
}

File: lib/rin/rhi/DescriptorAllocator.cpp
#include "rin/rhi/DescriptorAllocator.h"

#include "rin/rhi/GraphicsModule.h"

namespace rin::rhi
{

    vk::DescriptorPool DescriptorAllocator::GetPool()
    {
        vk::DescriptorPool pool{};

        if(!_readyPools.empty())
        {
            pool = *_readyPools.begin();
        }
        else
        {
            pool = CreatePool(_setsPerPool,_ratios);
            _setsPerPool = static_cast<uint32_t>(static_cast<float>(_setsPerPool) * 1.5f);
            if(_setsPerPool > 4092) _setsPerPool = 4092;
            _readyPools.emplace(pool);
        }
        return pool;
    }
    vk::DescriptorPool DescriptorAllocator::CreatePool(uint32_t sets, const std::vector<PoolSizeRatio>& ratios) const
    {
        std::vector<vk::DescriptorPoolSize> poolSizes{};
        poolSizes.reserve(ratios.size());
        for (auto &ratio : ratios)
        {
            poolSizes.emplace_back(ratio.type,static_cast<uint32_t>(ratio.ratio * static_cast<float>(sets)));
        } 
        vk::DescriptorPoolCreateInfo createInfo{_createFlags,sets,poolSizes};
        return _device.createDescriptorPool(createInfo);
    }
    DescriptorSet* DescriptorAllocator::Allocate(const vk::DescriptorPool& pool, const vk::DescriptorSetLayout& layout, const std::vector<uint32_t>& variableCounts)
    {
        if(!_sets.contains(pool))
        {
            _sets.emplace(pool,std::vector<DescriptorSet*>{});
        }
        
        vk::DescriptorSetAllocateInfo allocateInfo{pool,{layout}};

        DescriptorSet * set;
        if(!variableCounts.empty())
        {
            vk::DescriptorSetVariableDescriptorCountAllocateInfo variableInfo{variableCounts};
            allocateInfo.pNext = &variableInfo;
            set = new DescriptorSet(_device.allocateDescriptorSets(allocateInfo).at(0));
        }
        else
        {
            set = new DescriptorSet(_device.allocateDescriptorSets(allocateInfo).at(0));
        }
        _sets.at(pool).emplace_back(set);
        return set;
    }
    DescriptorAllocator::DescriptorAllocator(uint32_t maxSets, const std::vector<PoolSizeRatio>& ratios, const vk::DescriptorPoolCreateFlags& poolCreateFlags)
    {
        _ratios = ratios;
        _setsPerPool = maxSets;
        _device = GraphicsModule::Get()->GetDevice();
        _createFlags = poolCreateFlags;
    }
    DescriptorAllocator::~DescriptorAllocator()
    {
        DestroyPools();
    }
    DescriptorSet* DescriptorAllocator::Allocate(const vk::DescriptorSetLayout& layout, const std::vector<uint32_t>& variableCounts)
    {
        auto pool = GetPool();
        DescriptorSet * set{nullptr};
        try
        {
            set = Allocate(pool,layout,variableCounts);
        }
        catch(...)
        {
            _fullPools.emplace(pool);
            _readyPools.erase(pool);
            pool = GetPool();
            set = Allocate(pool,layout,variableCounts);
        }

        _readyPools.emplace(pool);

        return set;
    }
    void DescriptorAllocator::DestroyPools()
    {
        for (auto &pool : _fullPools)
        {
            for (const auto set : _sets.at(pool))
            {
                delete set;
            }
            _sets.erase(pool);
            _device.destroyDescriptorPool(pool);
        }

        _fullPools.clear();

        for (auto &pool : _readyPools)
        {
            for (const auto set : _sets.at(pool))
            {
                delete set;
            }
            _sets.erase(pool);
            _device.destroyDescriptorPool(pool);
        }

        _readyPools.clear();
    }
    
    void DescriptorAllocator::ResetPools()
    {
        for (auto &pool : _readyPools)
        {
            for (const auto set : _sets.at(pool))
            {
                delete set;
            }
            _sets.erase(pool);
            _device.resetDescriptorPool(pool);
        }

        for (auto &pool : _fullPools)
        {
            for (const auto set : _sets.at(pool))
            {
                delete set;
            }
            _sets.erase(pool);
            _device.resetDescriptorPool(pool);
            _readyPools.emplace(pool);
        }

        _fullPools.clear();
    }
}
